data/libs/Passengers.lua

```diff
@@ -0,0 +1,183 @@
+-- Copyright © 2008-2024 Pioneer Developers. See AUTHORS.txt for details
+-- Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+local Equipment = require 'Equipment'
+local EquipSet = require 'EquipSet'
+
+local utils = require 'utils'
+
+-- Module: Passengers
+--
+-- Helper utilities for working with passengers onboard a ship
+local Passengers = {}
+
+
+-- Function: CountOccupiedCabins
+--
+-- Return the number of occupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountOccupiedCabins(ship)
+	return ship["cabin_occupied_cap"] or 0
+end
+
+-- Function: CountFreeCabins
+--
+-- Return the number of unoccupied cabins present on the ship
+--
+---@param ship Ship
+---@return integer
+function Passengers.CountFreeCabins(ship)
+	return (ship["cabin_cap"] or 0) - (ship["cabin_occupied_cap"] or 0)
+end
+
+-- Function: GetOccupiedCabins
+--
+-- Return a list of currently occupied cabin equipment items
+--
+---@param ship Ship
+---@return Equipment.CabinType[]
+function Passengers.GetOccupiedCabins(ship)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetNumPassengers() > 0
+	end)
+end
+
+-- Function: GetFreeCabins
+--
+-- Return a list of currently free passenger cabins
+--
+---@param ship Ship
+---@param numPassengers integer?
+---@return Equipment.CabinType[]
+function Passengers.GetFreeCabins(ship, numPassengers)
+	local equipSet = ship:GetComponent("EquipSet")
+
+	return equipSet:GetInstalledWithFilter('Equipment.CabinType', function(equip)
+		return equip:GetFreeBerths() >= (numPassengers or 1)
+	end)
+end
+
+-- Function: CheckEmbarked
+--
+-- Validate how many of the given list of passengers are present on this ship
+--
+---@param ship Ship
+---@param passengers Character[]
```

Comment:
Table identity. Basically a pointer comparison in the Lua runtime. Two characters can be generated with the exact same names, faces, and seeds, but will still be different characters because they occupy separate regions of memory.

Reply:
Presumably this identity survives a load/store?  Might need a unique ID in future for more persistent characters.

Anyway, out -of-scope; closing. But thanks for the information.